### PR TITLE
Improve filtering of Lightning app logs

### PR DESCRIPTION
**Pull request by alecmerdler**
## What does this PR do?

Adds filtering of internal log messages using the delimiters
that separate different sections of the logs.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning_app/utilities/app_logs.py [hunk 1]
```diff
@@ -109,18 +143,20 @@ def _app_logs_reader(
...
 
-            if start_timestamp and log_event.timestamp > start_timestamp:
+            if component_logs_started[log_event.component_name] and not is_internal_log(log_event.message):
                 yield log_event
+                continue
+
+            section = LogSection.USER_RUN_FLOW if log_event.component_name == "flow" else LogSection.USER_RUN_WORK
+
+            if log_delimiter_begin_for(section) in log_event.message:
+                component_logs_started[log_event.component_name] = True
```
Comment (manskx) [inline src/lightning_app/utilities/app_logs.py:159]: If the app is restarted, will the logs contain the `delimiter_begin` multiple times ?

Comment (alecmerdler) [reply src/lightning_app/utilities/app_logs.py:159]: Right now, `lightning show logs` doesn't behave the same as the UI, which clears the logs after every app restart. We could do this in the CLI as well by passing `start` query parameter to the logs endpoint. We could include that in this PR or in a follow up.

**Comment by nicolai86** [comment]
can we add individual `lightning_app` files to our mypy config so we reduce the missing typing debt in that area over time?

**Comment by alecmerdler** [comment]
@nicolai86 That's a good idea. I'm less familiar with MyPy tooling, so someone else would be better to implement that.

File: src/lightning_app/utilities/app_logs.py (hunk 1 above)
Comment (manskx) [reply src/lightning_app/utilities/app_logs.py:159]: I think the logs should be cleared after every app stop/start, let's do it in a separate PR or this one if you prefer.

Maybe with an additional flag, we could show all logs in the future :D 
`lightning show logs --all`

File: src/lightning_app/utilities/app_logs.py [hunk 2]
```diff
@@ -40,6 +40,41 @@ def __gt__(self, other: "_LogEvent") -> bool:
         return self.timestamp > other.timestamp
 
 
+class LogSection:
+    """Delimiters used to separate sections of Lightning App logs."""
+
+    USER_RUN_FLOW = "USER_RUN_FLOW"
```
Comment (hhsecond) [inline src/lightning_app/utilities/app_logs.py:46]: We should probably show the user dependency installation as well

Comment (alecmerdler) [reply src/lightning_app/utilities/app_logs.py:46]: @hhsecond Getting the CLI to have complete parity with how the UI displays logs is a bigger task than this pull request was intended to cover.

File: src/lightning_app/utilities/app_logs.py (hunk 1 above)
Comment (nohalon) [reply src/lightning_app/utilities/app_logs.py:159]: I agree that the logs should get cleared after every app start/stop. can we add that to this PR

File: src/lightning_app/utilities/app_logs.py (hunk 2 above)
Comment (nohalon) [reply src/lightning_app/utilities/app_logs.py:46]: @tchaton

Comment (tchaton) [reply src/lightning_app/utilities/app_logs.py:46]: Hey, @alecmerdler We should do this on the platform side. It doesn't make sense to filter the logs with the private indications in an open-source project.

**Comment by nohalon** [comment]
Let's handle this kind of thing from the platform side. Chatted with @tchaton about it and I agree with him that this needs to be passed properly from the platform

File: src/lightning_app/utilities/app_logs.py (hunk 2 above)
Comment (alecmerdler) [reply src/lightning_app/utilities/app_logs.py:46]: @tchaton This logic is just copied from what we do in the UI, which any dedicated person could reverse-engineer. It's significantly more work to completely filter internal things from the external logs API and also have our nice app startup progress indicator. This was intended to be a :bicyclist: solution that would last us a few months until we had time to implement the next iteration.

**Pull request closed without merging**